File: src/io/github/knit_prg/kmc/misskey/MisskeyTimeline.java
package io.github.knit_prg.kmc.misskey;

import io.github.knit_prg.kmc.Gui;
import io.github.knit_prg.kmc.Settings;
import io.github.knit_prg.kmc.misskey.guiComponents.NoteCreator;

import java.awt.BorderLayout;
import java.awt.Container;
import java.awt.event.WindowAdapter;
import java.awt.event.WindowEvent;
import javax.swing.Box;
import javax.swing.BoxLayout;
import javax.swing.JPanel;

/**
 * Misskeyのタイムライン画面を示す
 *
 * @author Knit prg.
 * @since 0.1.0
 */
public final class MisskeyTimeline {

	/**
	 * 画面を開く
	 *
	 * @since 0.1.0
	 */
	public static void open() {
		Container contentPane = Gui.mainFrame.getContentPane();
		contentPane.removeAll();
		contentPane.setLayout(new BorderLayout());
		JPanel centerPanel = new JPanel();
		centerPanel.setLayout(new BoxLayout(centerPanel, BoxLayout.Y_AXIS));
		contentPane.add(centerPanel, BorderLayout.CENTER);
		centerPanel.add(new NoteCreator());
		centerPanel.add(Box.createHorizontalGlue());
		Streaming streaming = new Streaming(Settings.getInstance().getTokens().get(0), Streaming.Channel.LOCAL_TIME_LINE);
		streaming.connect();
		Gui.mainFrame.addWindowListener(new WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent e) {
				streaming.disconnect();
			}
		});
		contentPane.setVisible(false);
		contentPane.setVisible(true);
		contentPane.repaint();
	}
}

File: src/io/github/knit_prg/kmc/misskey/guiComponents/MisskeyTimeline.kt
package io.github.knit_prg.kmc.misskey.guiComponents

import com.fasterxml.jackson.databind.ObjectMapper
import io.github.knit_prg.kmc.Gui
import io.github.knit_prg.kmc.Settings
import io.github.knit_prg.kmc.misskey.MisskeyStreaming
import io.github.knit_prg.kmc.misskey.objects.MisskeyNote
import java.awt.BorderLayout
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import javax.swing.BoxLayout
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.ScrollPaneConstants

/**
 * Misskeyのタイムラインを示す
 *
 * @author Knit prg.
 * @since 0.1.0
 */
class MisskeyTimeline : JPanel() {

	val innerPanel = object : JPanel() {
		init {
			layout = BoxLayout(this, BoxLayout.Y_AXIS)
		}
	}

	val scrollPane = object : JScrollPane(innerPanel) {
		init {
			horizontalScrollBarPolicy = ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER
		}
	}

	/**
	 * つくる
	 *
	 * @since 0.1.0
	 */
	init {
		layout = BorderLayout()
		add(scrollPane, BorderLayout.CENTER)
		val streaming = MisskeyStreaming(
			Settings.instance.tokens[0],
			MisskeyStreaming.Channel.HYBRID_TIME_LINE
		)
		streaming.connect()
		streaming.onReceives.add { text ->
			val json = ObjectMapper().readTree(text.toString())
			when (json?.get("body")?.get("type")?.asText()) {
				"note" -> {
					json.get("body")?.get("body")?.let {
						innerPanel.add(NoteViewer(MisskeyNote(it)), 0)
						innerPanel.isVisible = false
						innerPanel.isVisible = true
					}
				}

				else -> {
					println(json)
				}
			}
		}
		Gui.mainFrame.addWindowListener(object : WindowAdapter() {
			override fun windowClosing(e: WindowEvent?) {
				streaming.disconnect()
			}
		})
	}
}